cli/azd/design-notes/environments.md

```diff
@@ -0,0 +1,230 @@
+```
+
+The `type` property of the `azd` metadata object denotes that this value should be set to an Azure location name (e.g. "westus2") and is used by `azd` to offer a "pick a location from this list of all Azure locations). This composes with other validation decorators that you can apply, for example `@allowed()` which can be used to restrict the set of locations offered to the user to pick from (for cases where a template uses services that are not deployed to all regions):
+
+```bicep
+@description('The Azure location to use for all resources.')
+@allowed([
+    'westus2'
+    'centralus'
+])
+@metadata({
+    azd: {
+        type: 'location'
+    }
+})
+param location string
+```
+
+To start, we support the following types:
+
+`location`: The value should be the short name of an Azure location. All locations are supported by default, but the exiting `@allowed` may restrict this set.
+`current-principal-id`: The value should be the id of the authenticated principal.
+
+Configuration values (including the environment name and id) are **not** stored in the `.env` file.
+
+Deployment outputs are also stored in the `config.json` file under the "outputs" key (a sibling of "properties") and updated on each deployment. They are written to the `.env` file as before, as well.
+
+The intermediate `.parameters.json` files are no longer written in the `.azure\<environment-name>` directory instead they are written in a temporary directory (and when we stop depending on `az` for deployments will not be persisted to disk)
+
+When creating names for deployments, we name them using the combination of the environment name and ID (e.g. `default-1539e0b8-ac32-440c-ae1f-d131512563dd`), if this would generate a deployment name which is too long, we truncate the environment name but keep the full id). This ensures the deployment names don't clash between two unrelated projects even if they have the same environment name (which will now be common, since we have a default value and no longer force the user into picking a new environment).
+
+We also need to understand the "target" of the deployment. Today, this is simply the subscription ID (since we only support subscription level deployments) and a location to store deployment metadata (which we infer from `AZURE_LOCATION`) but when we start to support resource level deployments this will change to be a subscription and resource group name. Terraform and Pulumi also support targeting multiple subscriptions in a single deployment operation (by creating multiple providers) so I imagine this target will become provider specific over time (I am guessing that for Pulumi and Terraform the target is more tied to a backend but I'm not 100% sure).
+
+For now, we'll introduce a `target` property:
+
+```json
+{
+  "id": "9fdfa455-2131-4faf-b2e3-8380aba97c96",
+  "name": "default",
+  "infra": {
+    "main": { 
+      "target": {
+        "type": "arm",
+        "subscriptionId": "faa080af-c1d8-40ad-9cce-e1a450ca5b57",
```

Comment:
We would have to design a gesture, but I agree that it important for us to do so.  I think we will need to think through this more as we start to think about different deployment backends.

We've imagined an `azd infra show` type command that lists information about your application's infrastructure (key resources, portal links, etc) - this could be a good place to surface this.

The way this is knowable today is by running `azd env get-values` and just knowing that the value of `AZURE_SUBSCRIPTION_ID` is what the deployment system targets.

Reply:
Added some comments about how to manage this as part of my last pass over the doc.